hazelcast/src/main/java/com/hazelcast/sql/impl/expression/datetime/Formatter.java

```diff
@@ -0,0 +1,1022 @@
+                PatternElement inferred = bracket == null && pair.bracket != null ? pair.bracket
+                        : pre && !(negative | pair.negative) && digits + pair.digits > 0
+                                && form != Form.Roman ? PatternElement.M : null;
+                if (inferred != null) {
+                    groups.add(offerSign, inferred);
+                }
+            }
+
+            void format(StringBuilder s, boolean negative, String integer, String fraction, int exponent,
+                        boolean overflow, DecimalFormatSymbols symbols) {
+                StringBuilder digits = overflow ? null : pre ? new StringBuilder(integer).reverse()
+                        : new StringBuilder(fraction);
+                if (!overflow) {
+                    // Append zeros to meet the minimum number of digits
+                    while (digits.length() < minDigits) {
+                        digits.append('0');
+                    }
+                    // Trim unnecessary zeros
+                    int length = digits.length();
+                    while (length > minDigits && digits.charAt(length - 1) == '0') {
+                        length--;
+                    }
+                    digits.setLength(length);
+                }
+                List<Object> parts = new ArrayList<>();
+                // d: Accumulate the count of processed digits in the mask
+                // f: Accumulate the filler space until a fixed element or the end of the mask is encountered
+                for (int i = 0, d = 0, f = 0; i < groups.size();) {
+                    Object g = groups.get(i);
+                    if (g instanceof Literal) {
+                        parts.add(((Literal) g).contents);
+                    } else if (g instanceof PatternElement) {
+                        PatternElement p = (PatternElement) g;
+                        if (p.isSign()) {
+                            char c = p.getSign(pre, negative);
+                            if (c != ' ') {
+                                parts.add(c);
+                            } else if (fillMode) {
+                                f++;
+                            }
+                        } else if (p == PatternElement.CR || p == PatternElement.C) {
+                            parts.add(symbols.getCurrencySymbol());
+                        } else if (p == PatternElement.TH || p == PatternElement.th) {
+                            if (integer != null && !integer.isEmpty()) {
+                                String th = getOrdinal(integer);
+                                parts.add(p == PatternElement.TH ? th.toUpperCase() : th);
+                            } else if (fillMode) {
+                                f += 2;
+                            }
+                        } else if (p == PatternElement.EEEE || p == PatternElement.eeee) {
+                            if (overflow || exponent != 0) {
```

Comment:
We must print the `"e+00"` also if the exponent is zero.

Reply:
We can print E+00 if the fill mode is active. Otherwise, skipping the exponent is more desirable and also consistent with other fillers.

Reply:
Consider this output:
```
VALUE
-----
1.500E+01
5.000
1.200E+02
1.861E+05
4.000
1.260E+02
```
versus:
```
VALUE
-----
1.500E+01
5.000E+00
1.200E+02
1.861E+05
4.000E+00
1.260E+02
```
I find the latter more desirable. So does PG and Oracle. Also on a calculator, if you enable scientific mode, you always see the exponent, even if it's zero. I find it very unusual to hide the exponent if it's 0.

Reply:
Consider this output:
```
VALUE
-----
1.500E+01
5.000
1.200E+02
1.861E+05
4.000
1.260E+02
```
versus:
```
VALUE
-----
1.500E+01
5.000E+00
1.200E+02
1.861E+05
4.000E+00
1.260E+02
```
I find the latter more desirable. So does PG and Oracle. Also on a calculator, if you enable scientific mode, you always see the exponent, even if it's zero. I find it very unusual to hide the exponent if it's 0.

Reply:
I agree the `g` pattern makes sense and might be useful, but now we don't have an alternative to the `e` pattern (I presume it prints the exponent unconditionally), which use useful too. In that case I'd follow what others do in this case.

Reply:
Currently, we don't let users to choose between `e`/`g` behavior. `EEEE` behaves like `e` in the padding mode, i.e. it unconditionally prints the exponent, and behaves like `g` in the fill mode, i.e. it only prints the non-zero exponent.

Reply:
It should behave as `e` in both mode.